astropy/cosmology/core.py

```diff
@@ -149,44 +175,127 @@ def format_spec(self):
         """String format specification."""
         return self._fmt
 
+    @property
+    def fixed(self):
+        """Whether the Parameter is fixed; true parameters are not."""
+        return self._fixed
+
+    # -------------------------------------------
+    # 'property' descriptor overrides
+
+    @property
+    def fget(self):
+        """Parameter value getter."""
+        return self._fget
+
+    def getter(self, fget):
+        """Make new Parameter with custom ``fget``.
+
+        Note: ``Parameter.getter`` must be the top-most descriptor decorator.
+
+        Parameters
+        ----------
+        fget : callable
+
+        Returns
+        -------
+        `~astropy.cosmology.Parameter`
+            Copy of this Parameter but with custom ``fget``.
+        """
+        return type(self)(fget=fget, fvalidate=self.fvalidate,
+                          doc=self.__doc__, fmt=self.format_spec,
+                          unit=self.unit, equivalencies=self.equivalencies,
+                          fixed=self.fixed)
+
     # -------------------------------------------
-    # descriptor
+    # descriptor methods
 
     def __get__(self, obj, objcls=None):
         # get from class
         if obj is None:
             return self
         # get from obj, allowing for custom ``getter``
-        if self.fget is None:  # default to private attr (diff from `property`)
+        if self._fget is None:  # default to private attr (diff from `property`)
             return getattr(obj, self._attr_name_private)
-        return self.fget(obj)
+        return self._fget(obj)
 
     def __set__(self, obj, value):
-        raise AttributeError("can't set attribute")
+        """Allows attribute setting once. Raises AttributeError subsequently."""
+        # raise error if setting 2nd time or has custom getter
+        # TODO! custom getter is currently a strong limitation on `validate`
+        #       because can't know if getter draws from ``_attr_name_private``
+        if hasattr(obj, self._attr_name_private) or self.fget is not None:
```

Comment:
> Note: another option would be to have a flag on the Cosmology instance on whether one is still in "initialization mode". But what you have here feels cleaner.

Agreed. I tried that. It got hacky.

Reply:
> Yet another option: lazyproperty stores the attribute under its own name in obj.__dict__ - since for a descriptor that is known to be safe. Would that make sense here too?

Let me time it all out again. I recall this implementation ended up being slow. Almost double the time.